Project-Aurora/Project-Aurora/Devices/Uniwill/KeyboardDevice.cs

```diff
@@ -0,0 +1,346 @@
+                    //    return true;
+                    //}
+                    //catch (Exception ex)
+                    //{
+                    //    Global.logger.Error("keyboard device, Exception! Message:" + ex);
+                    //}
+
+
+
+                   // Mark Initialized = FALSE
+                   isInitialized = false;
+                    return false;
+                }
+ 
+
+            return isInitialized;
+
+            }
+
+            // Handle Logon Event
+            //void SystemEvents_SessionSwitch(object sender, SessionSwitchEventArgs e)
+            //{
+            //    if (this.IsInitialized() && e.Reason.Equals(SessionSwitchReason.SessionUnlock))
+            //    { // Only Update when Logged In
+            //        this.SendColorsToKeyboard(true);
+            //    }
+            //}
+
+            public void Shutdown()
+            {
+                if (this.IsInitialized())
+                {
+                    if(CheckGCPower())
+                     keyboard?.release();
+
+                    bRefreshOnce = true;
+                    isInitialized = false;
+                   
+                }
+            }
+
+            public void Reset()
+            {
+                if (this.IsInitialized())
+                {
+                 if(CheckGCPower())
+                    keyboard?.release();
+
+                 bRefreshOnce = true;
+                 isInitialized = false;
+                
+                }
+                   
+            }
+
+            public bool Reconnect()
+            {
+                throw new NotImplementedException();
+            }
+
+            public bool IsInitialized()
+            {
+        
+               return isInitialized;
+            }
+
+            public bool IsConnected()
+            {
+                 return isInitialized;
+           }
+       
+        public bool UpdateDevice(Dictionary<DeviceKeys, Color> keyColors, DoWorkEventArgs e, bool forced = false) // Is this necessary?
+        {
+            if (e.Cancel) return false;
+
+            bool update_result = false;
+
+            watch.Restart();
+
+
+            //Alpha necessary for Global Brightness modifier
+            List<DeviceKeys> KeysList = new List<DeviceKeys>();
+
+            foreach (var key in keyColors.Keys) KeysList.Add(key);
+
+            foreach (var item in KeysList)
+            {
+                keyColors[item] = System.Drawing.Color.FromArgb(255, Utils.ColorUtils.MultiplyColorByScalar(keyColors[item],  (keyColors[item].A / 255.0D) * brightness));
+            }
+
+
+            keyboard?.SetEffect(0x32, 0x00, bRefreshOnce, keyColors, e);
+
+            bRefreshOnce = false;
+            watch.Stop();
+
+            lastUpdateTime = watch.ElapsedMilliseconds;
+
+            return update_result;
+        }
+
+        public bool UpdateDevice(DeviceColorComposition colorComposition, DoWorkEventArgs e, bool forced = false)
```

Comment:
This method is usually used as a timer proxy for the other UpdateColors method. if you don't need the other component of colorComposirtion, please consult the other device implementations for the usual usage. Repeating the same code in 2 methods doesnt do any good here.

Reply:
I will  combine to one method.